src/main/java/net/earthcomputer/clientcommands/command/WaypointCommand.java

```diff
@@ -0,0 +1,413 @@
+        compoundTag.getAllKeys().forEach(worldIdentifier -> {
+            CompoundTag worldWaypoints = compoundTag.getCompound(worldIdentifier);
+            waypoints.put(worldIdentifier, worldWaypoints.getAllKeys().stream()
+                .collect(Collectors.toMap(Function.identity(), name -> {
+                    CompoundTag waypoint = worldWaypoints.getCompound(name);
+                    BlockPos pos = NbtUtils.readBlockPos(waypoint, "pos").orElseThrow();
+                    ResourceKey<Level> dimension = Level.RESOURCE_KEY_CODEC.parse(new Dynamic<>(NbtOps.INSTANCE, waypoint.get("Dimension"))).resultOrPartial(LOGGER::error).orElseThrow();
+                    return new WaypointLocation(dimension, pos);
+                })));
+        });
+    }
+
+    private static Component formatCoordinates(BlockPos waypoint) {
+        return ComponentUtils.wrapInSquareBrackets(Component.literal(waypoint.toShortString())).withStyle(style -> style
+            .withColor(ChatFormatting.GREEN)
+            .withClickEvent(new ClickEvent(ClickEvent.Action.COPY_TO_CLIPBOARD, waypoint.getX() + " " + waypoint.getY() + " " + waypoint.getZ()))
+            .withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, Component.translatable("chat.copy.click")))
+        );
+    }
+
+    public static void renderWaypointLabels(GuiGraphics guiGraphics, DeltaTracker deltaTracker) {
+        String worldIdentifier = getWorldIdentifier(Minecraft.getInstance());
+        Map<String, WaypointLocation> waypoints = WaypointCommand.waypoints.get(worldIdentifier);
+        if (waypoints == null) {
+            return;
+        }
+
+        Minecraft minecraft = Minecraft.getInstance();
+        GameRenderer gameRenderer = minecraft.gameRenderer;
+        Camera camera = gameRenderer.getMainCamera();
+        Entity cameraEntity = camera.getEntity();
+        float partialTicks = deltaTracker.getGameTimeDeltaPartialTick(true);
+        double verticalFovRad = Math.toRadians(gameRenderer.getFov(camera, partialTicks, false));
+        Window window = minecraft.getWindow();
+        double aspectRatio = (double) window.getGuiScaledWidth() / window.getGuiScaledHeight();
+        double horizontalFovRad = 2 * Math.atan(Math.tan(verticalFovRad / 2) * aspectRatio);
+
+        Vec3 viewVector3 = cameraEntity.getViewVector(1.0f);
+        Vector2d viewVector = new Vector2d(viewVector3.x, viewVector3.z);
+        Vector2d position = new Vector2d(cameraEntity.getEyePosition().x, cameraEntity.getEyePosition().z);
+
+        List<WaypointLabelLocation> xPositions = new ArrayList<>();
+        waypoints.forEach((waypointName, waypoint) -> {
+            if (!waypoint.dimension().location().equals(minecraft.level.dimension().location())) {
+                return;
+            }
+
+            double distanceSquared = waypoint.location().distToCenterSqr(cameraEntity.position());
+            long distance = Math.round(Math.sqrt(distanceSquared));
```

Comment:
Is there a reason to call `distToCenterSqr` if you're immediately square rooting it? Is there not a `distanceToCenter` function?

Reply:
There is not.